pkg/informant/cgroup.go

```diff
@@ -0,0 +1,228 @@
+	return nil
+}
+
+// handleCgroupSignals is an internal function that handles "memory high" signals from the cgroup
+func (s *CgroupState) handleCgroupSignalsLoop(config CgroupConfig) {
+	// FIXME: the process might exit after freezing the cgroup but before thawing it. This isn't
+	// really feasible to fix here; the parent process should handle the "thaw if frozen" check on
+	// exit.
+	//
+	// FIXME: we should have "proper" error handling instead of just panicking. It's hard to
+	// determine what the correct behavior should be if a cgroup operation fails, though.
+	//
+	// FIXME: enforce a minimum wait period between requesting scale-ups (otherwise can be bad when
+	// memory usage is oscillating around memory.high but there aren't available resources)
+
+	for {
+		// Wait for a new signal
+		select {
+		case err := <-s.mgr.ErrCh:
+			panic(fmt.Errorf("Error listening for cgroup signals: %w", err))
+		case <-s.upscaleEventsRecvr.Recv():
+			klog.Infof("Received upscale event. Updating cgroup memory.high")
+			func() {
+				s.updateLock.Lock()
+				defer s.updateLock.Unlock()
+
+				if err := s.setMemoryHigh(); err != nil {
+					panic(fmt.Errorf("Error handling upscale signal: %w", err))
+				}
+			}()
+		case <-s.mgr.MemoryHighEvent.Recv():
+			// Double-check that there isn't currently an upscale event.
+			select {
+			case <-s.upscaleEventsRecvr.Recv():
+				klog.Warningf("Received memory high event, but also upscale event. Handling upscale and ignoring memory high")
+				func() {
+					s.updateLock.Lock()
+					defer s.updateLock.Unlock()
+
+					if err := s.setMemoryHigh(); err != nil {
+						panic(fmt.Errorf("Error handling upscale signal: %w", err))
+					}
+				}()
+			default:
+				func() {
+					s.updateLock.Lock()
+					defer s.updateLock.Unlock()
+
+					if err := s.handleMemoryHighEvent(config); err != nil {
+						panic(fmt.Errorf("Error handling memory high event: %w", err))
+					}
+				}()
+			}
+		}
+
+	}
+}
+
+// handleMemoryHighEvent performs the "freeze cgroup, request upscale, thaw cgroup" operation, in
+// response to a "memory high" event for the cgroup
+//
+// This method waits on s.agents.UpscaleEvents(), so incorrect behavior will occur if it's called at
+// the same time as anything else that waits on the upscale events. For that reason, both this
+// function and s.setMemoryHigh() are dispatched from within s.handleCgroupSignalsLoop().
+//
+// This method MUST be called while holding s.updateLock.
+func (s *CgroupState) handleMemoryHighEvent(config CgroupConfig) error {
+	klog.Infof("Received memory high event. Freezing cgroup")
+
+	// Immediately freeze the cgroup before doing anything else.
+	if err := s.mgr.Freeze(); err != nil {
+		return fmt.Errorf("Error freezing cgroup: %w", err)
+	}
+
+	startTime := time.Now()
+
+	// Start a timer for the maximum time we'll leave the cgroup frozen for:
+	maxWaitBeforeThaw := time.Millisecond * time.Duration(config.MaxUpscaleWaitMillis)
```

Comment:
I always prefer "num * <unit>" also because duration is just an int64 if your config variable is in int64 you *might* not need to convert. 

also I'd probably use `MS` over `Millis`

Reply:
Opened #35.